controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/ScaleOperationTask.java

```diff
@@ -83,31 +85,57 @@ public ScaleOperationTask(final StreamMetadataTasks streamMetadataTasks,
     }
 
     @VisibleForTesting
-    public CompletableFuture<EpochTransitionRecord> runScale(ScaleOpEvent scaleInput, boolean runOnlyIfStarted, OperationContext context, String delegationToken) { // called upon event read from requeststream
+    public CompletableFuture<Void> runScale(ScaleOpEvent scaleInput, boolean isManualScale, OperationContext context, String delegationToken) { // called upon event read from requeststream
         String scope = scaleInput.getScope();
         String stream = scaleInput.getStream();
-        // create epoch transition node (metadatastore.startScale)
+        // create epoch transition node (metadatastore.submitScale)
         // Note: if we crash before deleting epoch transition, then in rerun (retry) it will be rendered inconsistent
-        // and deleted in startScale method.
-        // if we crash before setting state to active, in rerun (retry) we will find epoch transition to be null and
-        // hence reset the state in startScale method before attempting to start scale in idempotent fashion.
-        return streamMetadataStore.startScale(scope, stream, scaleInput.getSegmentsToSeal(), scaleInput.getNewRanges(),
-                scaleInput.getScaleTime(), runOnlyIfStarted, context, executor)
-                .thenCompose(response -> streamMetadataStore.setState(scope, stream, State.SCALING, context, executor)
-                        .thenCompose(v -> streamMetadataStore.scaleCreateNewSegments(scope, stream, runOnlyIfStarted, context, executor))
+        // and deleted in submitScale method.
+        // if we crash before setting state to active, in rerun (retry) we will find epoch transition to be empty and
+        // hence reset the state here before attempting to start scale in idempotent fashion.
+        return streamMetadataStore.getVersionedState(scope, stream, context, executor)
+            .thenCompose(state -> streamMetadataStore.getEpochTransition(scope, stream, context, executor)
+                    .thenCompose(record -> {
+                        if (record.getObject().equals(EpochTransitionRecord.EMPTY)) {
+                            CompletableFuture<VersionedMetadata<State>> future = state.getObject().equals(State.SCALING) ?
+                                    streamMetadataStore.updateVersionedState(scope, stream, State.ACTIVE,
+                                            state, context, executor) : CompletableFuture.completedFuture(state);
+
+                            return future.thenCompose(updatedState -> {
+                                if (isManualScale) {
```

Comment:
Why do you wait so long to throw this? Can we do it earlier?

Reply:
replied against the other comment.